### PR TITLE
Add efficientnet v2 to keras.applications

### PR DESCRIPTION
**Pull request by sebastian-sz**
### Description

This PR adds EfficientNetV2 model architecture to `keras.applications`.

[Arxiv link](https://arxiv.org/abs/2104.00298)
[Reference implementation](https://github.com/google/automl/blob/master/efficientnetv2/effnetv2_model.py)

The code is heavily inspired by already existing code from `efficientnet.py` and `efficientnet_weight_update_util.py`. I tried to make it as similar to existing codebase as possible.

The models were rewritten using Keras functional API, and the weights were converted via modified weights conversion script.

### Motiviation
Adding EfficientNetV2 models to `keras.applications` has been discussed on Tensorflow's github:
https://github.com/tensorflow/tensorflow/issues/50054 

It was mentioned that this family of models would be welcomed in `keras.applications`, hence this contribution.

### Usage / Changes
The models could be used similarly to other models in `keras.applications`
```python
from keras.applications import efficientnet_v2

model = efficientnet_v2.EfficientNetV2_S()
```
### Todos / Known Issues
1. The output consistency was measured on a single panda image, from the original repository. It's worth noting that original repository performs cropping for some input sizes - I did not include this cropping here.
2. I did not use preprocessing layers, as I thought some users might experiment with different preprocessing. But they can be added.
3. Right now, I'm hosting the weights on GCP bucket but I won't be able to do this forever :(

### CODE REVIEW

**Comment by qlzh727** [comment]
Thanks for sending the PR, it will take us sometime to verify the details here. For the model weights, could you also attach them in the PR? (we can upload them to the keras-team gcs buckets, and then you can update the URL in the PR).

**Comment by sebastian-sz** [comment]
Thanks for the response. If I can help by answering any questions, let me know.

I'm having hard time attaching the weights on GitHub, so here is a Google Drive link: 
https://drive.google.com/file/d/1AYx_2zNyQGpYmcbsVeN1PewNUg999H8C/view?usp=sharing

**Comment by leondgarse** [comment]
It's good to have `efficientnetV2` directly usable form `keras.applications`. I have these model weights converted and uploaded in my own repo [leondgarse/keras_efficientnet_v2](https://github.com/leondgarse/keras_efficientnet_v2), including all `imagenet` / `imagenet-21k` / `imagenet21k-ft1k` pre-trained weights, and also `efficientnetv2-xl` weights. Hope this can help.

**Comment by innat** [comment]
@mingxingtan cc. @gbaned @qlzh727 
Could you please review or iterate over this [`efficientnet v2`](https://github.com/google/automl/tree/master/efficientnetv2) implementation? 
FYI, It's been implemented with functional API and not with the official way (subclassed API).

File: keras/applications/efficientnet_v2.py [hunk 1]
```diff
@@ -0,0 +1,1230 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (mingxingtan) [inline keras/applications/efficientnet_v2.py:1]: should it be 2021?

File: keras/applications/efficientnet_v2_weight_update_util.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Comment (mingxingtan) [inline keras/applications/efficientnet_v2_weight_update_util.py:1]: should it be 2021?

**Review by mingxingtan — COMMENTED** [review]
Thanks for adding it. Overall looks good to me!

**Comment by innat** [comment]
@qlzh727 @mattdangerw 
Any update on this PR?

File: keras/applications/efficientnet_v2.py [hunk 3]
```diff
@@ -0,0 +1,1230 @@
...
+EfficientNetV2_B1.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_B1')
+EfficientNetV2_B2.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_B2')
+EfficientNetV2_B3.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_B3')
+EfficientNetV2_S.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_S')
+EfficientNetV2_M.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_M')
+EfficientNetV2_L.__doc__ = BASE_DOCSTRING.format(name='EfficientNetV2_L')
+
+
+@keras_export('keras.applications.efficientnet_v2.preprocess_input')
+def preprocess_input(x, data_format=None):  # pylint: disable=unused-argument
```
Comment (fchollet) [inline keras/applications/efficientnet_v2.py:1217]: Our current policy is to *not* have a preprocessing function for new applications.

This function should just be pass-through, e.g.

```
def preprocess_input(x):
   return x
```

Meanwhile, the application itself should include a `Rescaling` layer as its first layer.

The goal being that applications should be able to process raw images, without the user having to care about what the correct preprocessing scheme to use is.

See e.g. 
https://github.com/keras-team/keras/blob/master/keras/applications/mobilenet_v3.py#L281-L282

Also include the following argument to make the Rescaling layer optional:

```
include_preprocessing: Boolean, whether to include the preprocessing
      layer (`Rescaling`) at the start of the network. Defaults to `True`.
```

https://github.com/keras-team/keras/blob/master/keras/applications/mobilenet_v3.py#L145

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR, this is great, let's ship it.

> The output consistency was measured on a single panda image, from the original repository. It's worth noting that original repository performs cropping for some input sizes - I did not include this cropping here.

Have you evaluated the model (with pretrained weights loaded) on the ImageNet validation set? What top-5 accuracy and accuracy do you get?

> Right now, I'm hosting the weights on GCP bucket but I won't be able to do this forever :(

We would download the files and re-host them in our own buckets as part of the merging process.

Also, keep in mind that once the application is submitted, you should update the keras.io docs with information about it: https://keras.io/api/applications/

**Comment by sebastian-sz** [comment]
@fchollet @mingxingtan 
Thank you for the review. In the nearest future I'll read the comments and fix the code.

File: keras/applications/applications_load_weight_test.py [hunk 4]
```diff
@@ -59,7 +60,12 @@
                      [efficientnet.EfficientNetB0, efficientnet.EfficientNetB1,
                       efficientnet.EfficientNetB2, efficientnet.EfficientNetB3,
                       efficientnet.EfficientNetB4, efficientnet.EfficientNetB5,
-                      efficientnet.EfficientNetB6, efficientnet.EfficientNetB7])
+                      efficientnet.EfficientNetB6, efficientnet.EfficientNetB7]),
+    'efficientnet_v2': (efficientnet_v2,
+                        [efficientnet_v2.EfficientNetV2_B0, efficientnet_v2.EfficientNetV2_B1,
+                        efficientnet_v2.EfficientNetV2_B2, efficientnet_v2.EfficientNetV2_B3,
+                        efficientnet_v2.EfficientNetV2_S, efficientnet_v2.EfficientNetV2_M,
+                        efficientnet_v2.EfficientNetV2_L])
```
[the remark this answers is not in the document]
Comment (sebastian-sz) [reply keras/applications/applications_load_weight_test.py:68]: If you prefer I can delete S, M, L variants. Let me know.

[the remark this answers is not in the document]
Comment (innat) [reply keras/applications/applications_load_weight_test.py:68]: @fchollet isn't it ok to keep all the v2 model's variants? In v1, it's included almost all. Also, in v2, the variants of `s`, `m`, `l` and `xl` have much higher top-1 accuracy compare to the rest (`bs`).

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks for the PR! I think Francois covered the main changes, but made a pass for style as well.

File: keras/applications/efficientnet_v2.py [hunk 5]
```diff
@@ -0,0 +1,1230 @@
...
+        to use as image input for the model.
+    input_shape: optional shape tuple, only to be specified
+        if `include_top` is False.
+        It should have exactly 3 inputs channels.
+    pooling: optional pooling mode for feature extraction
+        when `include_top` is `False`.
+        - `None` means that the output of the model will be
+            the 4D tensor output of the
+            last convolutional layer.
+        - `avg` means that global average pooling
```
[the remark this answers is not in the document]
Comment (sebastian-sz) [reply keras/applications/efficientnet_v2.py:849]: @mattdangerw I'm getting a bit confused with the backticks and string literals conventions. 
On line 836, you mentioned:
> `'imagenet' (pre-training on ImageNet),`
> surround string literal with backticks

but here, you mention that string literal should be surrounded by single quote `'avg'`.

1. Should this be surrounded by single quote?
2. Should I also change None and max parameters to be surrounded with single quote?

File: keras/applications/efficientnet_v2.py [hunk 6]
```diff
@@ -0,0 +1,1230 @@
...
+    if 0 < se_ratio <= 1:
+      filters_se = max(1, int(input_filters * se_ratio))
+      se = layers.GlobalAveragePooling2D(name=name + "se_squeeze")(x)
+      if bn_axis == 1:
+          se_shape = (filters, 1, 1)
+      else:
+          se_shape = (1, 1, filters)
+      se = layers.Reshape(se_shape, name=name + "se_reshape")(se)
+
+      # Warning. Is local pooling always false?
```
[the remark this answers is not in the document]
Comment (sebastian-sz) [reply keras/applications/efficientnet_v2.py:658]: Sorry, I forgot to delete this. 

As far as I know `local_pooling` is always false. You can check my analysis workflow [here](https://colab.research.google.com/drive/1bdyTwSc5raoWY2-9obm6oMAzvsm87i19?usp=sharing).

File: keras/applications/efficientnet_v2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras/applications/efficientnet_v2.py:849]: Oh sorry, I meant add the quotes along with the backticks here as well. It is just that github is grabbing my backticks and rendering them as markdown.

A string literal should be surrounded by backticks and quoted. None should be surrounded by backticks and not quoted. So \`'avg'\`, \`'max'\` and \`None\`.

Basically anything inside the backticks should be valid python you could pass for the argument in question.

File: keras/applications/efficientnet_v2.py [hunk 7]
```diff
@@ -0,0 +1,1230 @@
...
+      # The first block needs to take care of stride and filter size increase.
+      if j > 0:
+        args["strides"] = 1
+        args["input_filters"] = args["output_filters"]
+      
+      x = block(
+        inputs=x,
+        activation=activation,
+        bn_momentum=bn_momentum,
+        survival_probability=drop_connect_rate * b / blocks,  # Is this ok? Copied from V1.
```
[the remark this answers is not in the document]
Comment (sebastian-sz) [reply keras/applications/efficientnet_v2.py:962]: @fchollet 
I checked again and It should be ok.

V1 code:
```python
if survival_prob:
  drop_rate = 1.0 - survival_prob
  survival_prob = 1.0 - drop_rate * float(idx) / len(self._blocks)
  logging.info('block_%s survival_prob: %s', idx, survival_prob)
```

V2 equivalent:
```python
if survival_prob:
  drop_rate = 1.0 - survival_prob
  survival_prob = 1.0 - drop_rate * float(idx) / len(self._blocks)
  logging.info('block_%s survival_prob: %s', idx, survival_prob)
```

Sources:
V1: https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/efficientnet_model.py#L656
V2: https://github.com/google/automl/blob/master/efficientnetv2/effnetv2_model.py#L617

File: keras/applications/efficientnet_v2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebastian-sz) [reply keras/applications/efficientnet_v2.py:849]: Understood, thanks!

If you don't mind me asking one more question: 
Should I indent my code everywhere with 2 spaces instead of 4?

**Comment by sebastian-sz** [comment]
I updated the PR based on the comments. Let me know if this is ok.

### Model evaluation
> Have you evaluated the model (with pretrained weights loaded) on the ImageNet validation set? What top-5 accuracy and accuracy do you get?

Sadly no. I don't have the data and compute to run it, but I can help setting up the scripts.

### Weights hosting
> We would download the files and re-host them in our own buckets as part of the merging process.

I had to remove the weights from my end. Here is the [Google Drive link](https://drive.google.com/file/d/1UCtqTFQ5G-eg1LoK08qxTq3t-O-h0Sib/view?usp=sharing) with updated weights.

### Model variants.
> Do we really need all these models? Why not just the Bs (like for v1)?

Personally I wouldn't mind `S/M/L` variants but if you want them deleted I will remove them.
How should I proceed? Delete `S/M/L` variants, leave as is, or add `XL` to all variants?

**Comment by sebastian-sz** [comment]
@mattdangerw 
Sorry to interrupt, I can see that the PR is merged, but what about the weights upload and download path?

**Comment by mattdangerw** [comment]
@sebastian-sz I update the code during submission to point to `"https://storage.googleapis.com/tensorflow/keras-applications/efficientnet_v2/"`. Which has the weights from the google drive folder you shared. Updating those weights (and hashes) will need to be done by someone on the Keras team with write access to the bucket.

**Comment by sebastian-sz** [comment]
@mattdangerw Got it, thanks for the clarification!

**Comment by innat** [comment]
@sebastian-sz I've noticed that EffNetV2-XL is released. Will it also be included here? It's the big model I guess.

**Comment by sebastian-sz** [comment]
> @sebastian-sz I've noticed that EffNetV2-XL is released. Will it also be included here? It's the big model I guess.

@innat 
This is merged so I guess it can be added in a separate PR. 
The model is so big that there are two problems:
1. Current naming convention will fail due to huge number of layers. One needs to change:
from 
`name="block{}{}_".format(i + 1, chr(j + 97)),`
to something like
`name=f"block{i+1}-{j+1:02d}_",` 

The weights should not need to be reuploaded, but XL variant must be converted.

2. RecursionError
The model has so many layers that creating it directly:
```python
def EfficientNetV2XL(
    include_top=True,
    weights="imagenet",
    input_tensor=None,
    input_shape=None,
    pooling=None,
    classes=1000,
    classifier_activation="softmax",
):
    """Create EfficientNetV2 XL variant."""

    return EfficientNetV2(
        width_coefficient=1.0,
        depth_coefficient=1.0,
        default_size=512,
        model_name="efficientnetv2-xl",
        # (...)
    )
```
will throw `RecursionError`. One can increase recursion limit:
```python
current_limit = sys.getrecursionlimit()
    target_limit = 2000
    if current_limit < target_limit:
        sys.setrecursionlimit(target_limit)
```

but I don't know if it's a bug or there is a better way to handle this.

**Comment by innat** [comment]
@sebastian-sz Are [these ](https://github.com/sebastian-sz/efficientnet-v2-keras/releases)the exact weights file that will be used in `tf.keras.applicaitons` API?

**Comment by sebastian-sz** [comment]
@innat 

### About the weights
Very similar - the model's architecture is the same apart for few code differences - but not exactly the same.  
The differences are:
* weights in the link don't have preprocessing layer in the model (although this should not be an issue)
* different block naming - weights in the link use the one I mentioned
* weights in the link do not use EMA for easier comparison. In this PR I included EMA as I've heard somewhere they are preffered and even slightly better.
* Slightly different weight conversion script: for this PR i used [this script](https://gist.github.com/sebastian-sz/dc349b933a3b9bb1906773e1608c3acb). 

### About the issue
They do not have this Issue in the original repo as they use subclassing API. I guess in subclassing API there are fewer loops as they are only stacking blocks (e.g. STEM might be treated as a single layer). Functional API (this PR + my repo) is stacking raw layers (e. g. conv2D) layer after layer. 

I will open the Issue on Keras repo with an example and see what the devs say.
